### DISCUSSION TITLE
Pitch Detection

### DISCUSSION THREAD

**Discussion by eccessivo75** [pschatzmann/arduino-audio-tools#1005] [Q&A] [answered]
Hello,
I was trying to use the FFT To detect the pitch of a note played on a guitar, or a piano or a clarinet. My hardware is the espressif Lyrat v 4.3
RealFFT, and  ESP32FFT are very well implemented  and work very well with a sine, but, in my case, they didn't perform well when used to detect the pitch of a  musical instruments. The problem seems to be related to the harmonics that sometimes are louder than the fundamental.
Browsing internet i found a couple of projects that using a different approach called " autocorrelation" that seems to be very promising here is a link https://www.instructables.com/Reliable-Frequency-Detection-Using-DSP-Techniques/
Question:  Is there anyone has successfully implemented this algorithm with audiotools, or knows about an implementation?
Which is the best way to get the audiotools stream to the pitch detector?
Thanks

**Comment by pschatzmann** [comment, marked as answer]
I would doubt that something as primitive as this would give any good results. The big advantage of FFT is that it gives your the result over the full spectrum and not a single number: so you will get an see the harmonics and it is up to you what you make out if it. 

By just counting the zero crossings this gets complexly lost...

But feel free to extend my framework: I will be more then happy to accept a pull request for this.

I have the feeling that this discussion might lead to some better result that what you are proposing: https://github.com/pschatzmann/arduino-audio-tools/discussions/363

**Comment by pschatzmann** [comment]
I did a quick test of your proposed solution: The results of the effective frequencies vs the determined frequencies is looking pretty bad...

**Reply by eccessivo75** [reply]
Hello,
Thank you for testing on your side. Is the test code shared also on Github? If yes, i can double check using the same scenario and compare the performances with the FFT.
The project that you linked looks very interesting, it is something  that i found desio have.
The test i made with ESP32fft and Real FTT produced an unexpected output ant not very usable.
For lengths lower than 1024  i get, wrong frequencies and , weirdly, playing contiguous frets, produce the same leading frequency reading, even if  i am   increasing the frequency of half a tone each step. Everything improves  by increasing the lenght. With a lenght of 8192 readings are quinte good but it is not practical.
During all the process i check the audio at the board  output by using my earphones and the quality is ok. In the next  minutes, i will use a function generato to inject a sinewave  and verify the readings to avoid any hardware issue

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/sandbox/streams-generator-freq/streams-generator-freq.ino

There is nothing wrong with having big length: You can also try to optimize the sampling rate: 16000 should be good enough if you want to go as high as B8.

**Reply by eccessivo75** [reply]
Hi,
Thank you for the link , i was aware of that but i wanted  to use an external function generator to take care also of the physical input stage (i  have a FET preamp as impedance matching stage).
I was able to confirm that on the lower frequencies , short FFT length doesn't work , increasing the length give good results as predicted.
To Keep the FFT fast i decrease the sampling rate. I look forward for any development on the discussion https://github.com/pschatzmann/arduino-audio-tools/discussions/363.
Thank you
